HTTP2 proactive GOAWAY on drain-start

murray-stripe opened the pull request:
Signed-off-by: John Murray <murray@stripe.com>

Commit Message:
Change the default behavior of draining from a pull-based to a push-based model, proactively notifying interested parties when a drain sequence has begun. For low-volume listeners/connections, this allows for a graceful termination by proactively sending GOAWAY signals. This change resolves #14350.

This change relies the concept of "drain trees" (introduced in #17026), where the drain-manager exists as either a parent or child drain-manager. Allowing drain actions to cascade downward and also allowing for independent "sub-tree" draining (e.g. draining a specific Listener or specific FilterChain).

Additional Description:
Risk Level: moderate
Testing: Additional test coverage added for the HTTP ConnectionManagerImpl as well as the DrainManagerImpl. Many tests updated to reflect changes in behavior (push vs pull).
Docs Changes: Updated comments on DrainStrategy to convey some specifics of timing in how draining is initiated.
Release Notes:
Fixes #14350 

<--
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]
-->

antoniovicente commented on test/server/drain_manager_impl_test.cc:
```diff
@@ -151,6 +155,42 @@ TEST_P(DrainManagerImplTest, DrainDeadlineProbability) {
+  for (auto i = 0; i < 20; i++) {
+    auto& cb = cbs[i];
+    if (drain_gradually) {
+      auto step = 1000 / 20;
```
Would be good to cover the following test cases:
- number of callbacks does not evenly divide 1000msec, example: 30 callbacks
- number of callbacks is larger than 1000, so the step returned by this computation is 0.

murray-stripe replied:
- [ ] Add additional tests

antoniovicente commented on source/common/common/callback_impl.h:
```diff
@@ -81,5 +84,107 @@ template <typename... CallbackArgs> class CallbackManager {
+   * Run all registered callbacks.
+   * NOTE: This code is currently safe if a callback deletes ITSELF from within a callback. It is
+   *       not safe if a callback deletes other callbacks. If that is required the code will need
+   *       to change (specifically, it will crash if the next callback in the list is deleted).
```
Are there tests that cover this removal while iterating behavior?

I think that `remove` attempts to take lock_ despite remaining held while the callback executes.  IIRC the locks used here are not re-entrant, attempts to take a lock that is already held will deadlock.

murray-stripe replied:
- [ ] Add test for removal while iterating.

murray-stripe replied:
You are correct that the lock is not re-entrant and rather than trying to determine a clever way of not needing a lock, I've simply documented that removal is not a supported operation _within_ the callback. Let me know if this is an acceptable trade-off.

murray-stripe replied:
No longer true after refactor of callback manager.

antoniovicente reviewed (commented):
Thanks for the improvements to the drain sequence.  Here's some early feedback on your WIP changes.

murray-stripe commented on source/common/common/callback_impl.h:
```diff
@@ -81,5 +84,107 @@ template <typename... CallbackArgs> class CallbackManager {
+ * Utility class for managing callbacks. Thread-safe.
+ * @see CallbackManager
+ */
+template <typename... CallbackArgs> class ThreadSafeCallbackManager {
```
🤔  Hmm... the drain manager is the central object that is currently shared across threads, so my intention was to create a version of the `CallbackManager` that was thread-safe so that it could be used within the shared drain-manager object.

Open to another approach, but currently not sure how to go about it.

murray-stripe replied:
So for the moment I've worked around this limitation by wrapping the callback body in a dispaterhcer `post()` function which is thread-safe.

murray-stripe replied:
Resolving this thread post-refactor as they are no longer relevant.

murray-stripe commented on source/common/common/callback_impl.h:
```diff
@@ -81,5 +84,107 @@ template <typename... CallbackArgs> class CallbackManager {
+   *       to change (specifically, it will crash if the next callback in the list is deleted).
+   * @param args supplies the callback arguments.
+   */
+  void runCallbacks(CallbackArgs... args) {
```
I don't see any reason (generically) that this couldn't be a valid use-case. For the specific case of the drain-manager, I would say that it could make sense to release the callbacks after invoking them, but the clients own the handles. This fits the model of other uses of the CallbackManager I believe.

murray-stripe commented on source/common/common/callback_impl.h:
```diff
@@ -81,5 +84,107 @@ template <typename... CallbackArgs> class CallbackManager {
+  size_t size() const noexcept {
+    Thread::LockGuard lock(lock_);
+    return callbacks_.size();
+  }
```
- [ ] add assert

murray-stripe commented on source/common/common/callback_impl.h:
```diff
@@ -81,5 +84,107 @@ template <typename... CallbackArgs> class CallbackManager {
+  mutable Thread::MutexBasicLockable lock_;
+
+  ABSL_GUARDED_BY(lock_)
+  std::list<CallbackHolder*> callbacks_;
```
- [ ] formatting fix

murray-stripe commented on source/server/drain_manager_impl.cc:
```diff
@@ -52,15 +52,46 @@ bool DrainManagerImpl::drainClose() const {
+  // Call registered on-drain callbacks - immediately
+  if (server_.options().drainStrategy() == Server::DrainStrategy::Immediate) {
+    std::chrono::milliseconds no_delay{0};
+    cbs_.runCallbacks(no_delay);
```
- [ ] unify code-usage

murray-stripe commented on source/server/drain_manager_impl.cc:
```diff
@@ -52,15 +52,46 @@ bool DrainManagerImpl::drainClose() const {
+
+  // Call registered on-drain callbacks - with gradual delays
+  // Note: This will distribute drain events in the first 1/4th of the drain window
+  //       to ensure that we do not initiate draining with enough time for graceful
```
- [ ] fix comment wording

murray-stripe commented on source/server/drain_manager_impl.cc:
```diff
@@ -52,15 +52,46 @@ bool DrainManagerImpl::drainClose() const {
+  ASSERT(server_.options().drainTime() >= remaining_time);
+
+  const auto time_step =
+      std::chrono::duration_cast<std::chrono::milliseconds>(remaining_time) / cbs_.size() / 4;
```
Good point! I wasn't entirely satisfied with how this part landed and I think this is a much better approach.

- [ ] Fix drain-duration timing

murray-stripe commented on source/server/drain_manager_impl.cc:
```diff
@@ -52,15 +52,46 @@ bool DrainManagerImpl::drainClose() const {
+      [&](Common::ThreadSafeCallbackManager<std::chrono::milliseconds>::Callback cb) {
+        cb(time_step * step_count);
+        step_count++;
+      });
```
Hmmm... I can't imagine a scenario in which a drain event racing with callback registration is an appropriate ordering of events, but I'm not certain it isn't something that can happen. I don't think I have a strong preference for either:

- scheduling it for the near future
- scheduling it immediately
- disallowing it

I do agree that we should do _something_ though and document the decision. I'm leaning towards one of the scheduling options. thoughts?

antoniovicente replied:
I agree, let's pick one of these options and do it.  I'm in favor of the scheduling options.

murray-stripe replied:
This has been addressed.

murray-stripe commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -3421,18 +3420,23 @@ TEST_F(HttpConnectionManagerImplTest, DrainCloseRaceWithClose) {
   Event::MockTimer* drain_timer = setUpTimer();
   EXPECT_CALL(*drain_timer, enableTimer(_, _));
   expectOnDestroy();
+
```
- [ ] Add Http2 tests to verify correct GOAWAY behavior

murray-stripe replied:
I've added a couple of additional tests. One to verify that a GOAWAY is sent when the callback fires. The second to validate proper behavior if the callback fires after the conn-manager has entered into an initial draining state (callback should be a no-op).

murray-stripe commented on test/server/drain_manager_impl_test.cc:
```diff
@@ -151,6 +155,42 @@ TEST_P(DrainManagerImplTest, DrainDeadlineProbability) {
+            EXPECT_LT(delay.count(), 1001);
+
+            // Validate that our wait times are spread out (within some small error)
+            EXPECT_EQ(delay.count(), i * step);
```
Ahh... it _was_ fuzzy at one point. Adding a task to clean-up, but will likely become true again as I change the delay-time bits.

- [ ] Either fix comment or ensure it's true :-)

murray-stripe commented on source/server/drain_manager_impl.cc:
```diff
@@ -52,15 +52,50 @@ bool DrainManagerImpl::drainClose() const {
+  }
+
+  // Call registered on-drain callbacks - with gradual delays
+  // Note: This will distribute drain events in the first 1/4th of the drain window
```
ah, good idea. For now I'll document this as part of the config API. If the PR starts shaping up and the core bits are looking good, I may circle back around to add some additional configuration inputs.

- [ ] Document 1/4 behavior in config APIs

murray-stripe reviewed (commented):
@antoniovicente  Thank you for the review!! 😄 

I've added replies to all of your comments to either add a follow-up action (I'll be working through these) or to answer a question or start a discussion.

antoniovicente commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -141,6 +142,24 @@ void ConnectionManagerImpl::initializeReadFilterCallbacks(Network::ReadFilterCal
+        // body with `post()` to make sure it is executing on this thread.
+        read_callbacks_->connection().dispatcher().post([this, drain_delay]() -> void {
+          start_drain_timer_ =
+              read_callbacks_->connection().dispatcher().createTimer([this]() -> void {
```
I think there are still a few possible issues:
- read_callbacks_ is read across threads without synchronization, so the execution of the drain close cb could race with the read_callback termination sequence and call to removeOnDrainCloseCb
- related to the previous issue, what guarantees that read_callbacks_ is not deleted between the time the callback is posted and when it executes?

I think that we need a mechanism to track drain callbacks per worker and a way to propagate drains from the process wide object to the tracker per worker.  Other possible alternatives may include use of weak pointers or similar mechanisms to determine if read_callbacks_ is still live when attempting to execute the associated callback.  Also, the dispatcher associated with the read_callbacks_ should be captured explicitly so there's no cross-thread read of the read_callback_ data members.

murray-stripe replied:
@antoniovicente 
I agree with you that it would be best to manage things per-worker and I want to bake this abstraction into the thread-safe version of the callback manager so that it can be re-used safely (and just feels like a proper encapsulation of scope). 

I'm considering the following approach, but want to get your take on it. `ThreadSafeCallbackManager` would change the `add` function as:

```cpp
ThreadSafeCallbackHandlePtr add(Event::Dispatcher&, Callback); 
```

The `ThreadSafeCallbackHandlePtr` would be a `shared_ptr` and a `weak_ptr` would be retained along with the `Disptacher`. The `runCallbacksWith` function would look roughly the same from an API perspective as it does now but internally will use the `Dispatcher` to `post()` on the thread that registered the callbacks.  

__Deletion__ \
Since the returned callback-handle is now a `shared_ptr`, we don't have to perform any cleanup within the callback manager. Instead we can trim callbacks that have expired during normal operations of the class such as getting the size() or executing the callbacks. 

Locks will still be required when updating the central list, but since all add/remove operations would now be local to the class, I believe this would remove any sort of re-entrant locking behaviors.

murray-stripe replied:
I've implemented the refactor that I discussed above so resolving this conversation for now for re-review.

murray-stripe commented:
I'm still working through some of the testing (both in adding additional coverage from feedback and resolving failing tests due to changs) but I have covered drain-manager and callback-impl tests with the latest set of changes.

murray-stripe commented:
During my testing I realized that I missed the use-case of xDS updates and filter-chain draining. Looking through this code was a hierarchy of `DrainDecision` interfaces/implementations. Moving from a pull model to a push-model didn't fit this mode well as it stood.

I've updated the drain-manager to allow for "child" drain-managers that can inherit the parent's drain state, but also separately drain - effectively creating "drain trees" that mirror the prior push model. I'll need to add some additional test coverage for the drain-manager in this area and I'll also be working through some other failing tests.

murray-stripe commented:
@antoniovicente I am at a point where I would appreciate your feedback, but also understand if you'd like to wait until I've worked through the remaining test cleanup. Since your last review I've made some significant changes to the `ThreadSafeCallbackManager` as well as the `DrainManager` and some of the xDS classes use of them (see above comment). The remaining tasks on my plate are:

- [ ] Resolve all failing tests due to changes in drain-behavior (xds tests, conn-manager-impl, anything making use of `MockDrainManager`, etc.)
- [ ] Add additional coverage for drain-manager children

antoniovicente commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -460,6 +481,25 @@ void ConnectionManagerImpl::doConnectionClose(
+void ConnectionManagerImpl::createStartDrainTimer(std::chrono::milliseconds drain_delay) {
+  if (!codec_) {
+    stats_.named_.downstream_cx_drain_close_.inc();
+    doConnectionClose(Network::ConnectionCloseType::FlushWrite, absl::nullopt, "");
```
A burst of connection closes for connections with no active codec could result in a burst of work since the closes won't be spread over the close interval.

murray-stripe replied:
Ah, good catch. I've updated this so that the same timer is applied to either calling `doConnectionClose()` or `startDrainSequence()`.

murray-stripe replied:
Ah, this was changed and then changed back. A separate conversation in [this comment thread](https://github.com/envoyproxy/envoy/pull/16201#discussion_r672610649). I believe the two comments at two different times led us to two different conclusions. 

Happy to hear additional feedback on which might be a good approach to take here.

antoniovicente commented on source/server/drain_manager_impl.cc:
```diff
@@ -14,7 +16,34 @@ namespace Server {
+  child_drain_cbs_.push_back(addOnDrainCloseCb(
+      dispatcher,
+      [child_ = std::weak_ptr<DrainManagerImpl>(child)](std::chrono::milliseconds) mutable {
+        // disregard the specified delay time and kick off the child drain immediately
```
It seems that we're not spreading events when processing in child managers.  This will result in bursts of event callbacks when the child manager is notified.

murray-stripe replied:
Correct that child managers are notified immediately, but the same delay logic will be applied according to each child's configuration in the following `startDrainSequence` call. Essentially the draining happens at the leaf's of the tree and I would not expect an exceedingly large number of DrainManagers in the tree.

antoniovicente commented on include/envoy/server/drain_manager.h:
```diff
@@ -3,17 +3,35 @@
+   * be used to enact local draining.
+   *
+   * @param dispatcher Dispatcher for the current thread in which the new child drain-manager will
+   * exist.
```
Would be useful to expand this comment about intended use of child managers and how their behavior differs from their parent, e.i. ignores the drain time argument

murray-stripe replied:
Comment expanded.

antoniovicente commented on source/server/drain_manager_impl.cc:
```diff
@@ -14,7 +16,34 @@ namespace Server {
+  auto child = std::make_shared<DrainManagerImpl>(server_, drain_type, dispatcher);
+
+  // Wire up the child so that when the parent starts draining, the child also sees the state-change
+  child_drain_cbs_.push_back(addOnDrainCloseCb(
```
Is it possible for child_drain_cbs_ to grow very large over the lifetime of the Envoy process?  Possible example: a proxy that receives periodic config updates that add and remove listeners; each listener seem to have its own drain close cb, so the listener removals would result in an ever increasing list of expired child_drain_cbs_

murray-stripe replied:
It is definitely possible, especially for the main (server-owned) drain-manager. I'll update to address this.

antoniovicente commented on source/server/listener_impl.cc:
```diff
@@ -281,8 +283,7 @@ ListenerImpl::ListenerImpl(const envoy::config::listener::v3::Listener& config,
       continue_on_listener_filters_timeout_(config.continue_on_listener_filters_timeout()),
       listener_factory_context_(std::make_shared<PerListenerFactoryContextImpl>(
-          parent.server_, validation_visitor_, config, this, *this,
-          parent.factory_.createDrainManager(config.drain_type()))),
```
It seems to me that we could keep this old interface by calling `parent_.server_.drainManager().createChildManager(parent_.server_.dispatcher()` here.

Why should we prefer to pass in drain_type instead?

murray-stripe replied:
I think this was one of the details that changes for reasons during my development that is no longer relevant now. Will revert to maintain the original (or closer to) interface.

- [ ] Keep previous interface

mattklein123 replied:
Do we still need the `createDrainManager` on the factory?

murray-stripe replied:
Good call-out. We don't need this anymore and I've removed it.

antoniovicente reviewed (commented):
I think it looks better with regard to thread safety.  My remaining worries is about thundering herds of posted callbacks when a drain event is triggered.  Could use more info about the intended use of child managers.

/wait

murray-stripe commented on source/common/common/callback_impl.h:
```diff
@@ -81,5 +86,134 @@ template <typename... CallbackArgs> class CallbackManager {
+          still_alive_(parent.still_alive_) {}
+
+    ~CallbackHolder() override {
+      // schedule the removal on the parent thread to avoid cross-thread access or lock contention
```
I'm okay if post takes a lock because it means that we can retain behavior that exists in the non-threadsafe version of the `CallbackManager` that allows callbacks to de-register themselves within the callback. I made this change and took advantage in the HCM to do exactly this.

murray-stripe replied:
Moving discussion to new PR ([link to comment](https://github.com/envoyproxy/envoy/pull/17026/files?file-filters%5B%5D=.cc#r653703702)).

murray-stripe commented on source/common/common/callback_impl.h:
```diff
@@ -81,5 +86,134 @@ template <typename... CallbackArgs> class CallbackManager {
+      }
+
+      auto args = run_with();
+      cb_dispatcher.post([cb = cb, args = args] {
```
It's possible that I've missed something in my mental model, so I'll try to explain my mental model below because I don't believe what you're explaining is accurate, but please do correct me where I'm wrong.

---

The on-drain callback is registered per `Http::ConnectionManagerImpl` (HCM) instance. There is only one HCM instance per filter-chain, per-thread. In my current experience operating our mesh, the biggest I've seen is a 32 thread Envoy running 2 listeners each with 1 HTTP connection-manager defined in a single filter-chain. These machines see 10's of thousands of concurrent connections, but given this change I would only expect to see 64 registered drain callbacks (32 threads * 2 filter-chains * 1 HCM). 

Of course I can imagine running with a larger concurrency and configuring many more HTTP connection managers (and filter chains), but I am not connecting what sounds like a concern in the number of registered (and executed) callbacks and the number of open connections.

antoniovicente replied:
My understanding is that there is one FilterManagerImpl per Network::ConnectionImpl.  ConnectionManagerImpl per FilterManagerImpl also implies that there's one per downstream connection.  I do find this somewhat confusing and counter intuitive.

I think that code like this https://github.com/envoyproxy/envoy/blob/4d2e0187d2dd4136c57c0f47cf5a209bcc13bda4/source/common/http/conn_manager_impl.cc#L812 which goes from ConnectionManagerImpl instance to Network::Connection instance supports my impression about the relationships between these objects and number of HCMs per Envoy process.

murray-stripe replied:
Ahhh okay. This makes a lot more sense with the initial concerns then. Thank you for helping me form the right mental model! The comments on the original issue regarding batching also make more sense now. I'll spend some time figuring out how to batch these and/or avoid so many cross-thread calls.

murray-stripe replied:
This has been addressed in the current iteration, making use of TLS to reduce the cross-thread interaction and I have also made callback registration a _non-thread-safe_ operation to simplify the implementation.

murray-stripe commented on include/envoy/server/drain_manager.h:
```diff
@@ -3,17 +3,35 @@
+  virtual DrainManagerSharedPtr
+  createChildManager(Event::Dispatcher& dispatcher,
+                     envoy::config::listener::v3::Listener::DrainType drain_type) PURE;
+  virtual DrainManagerSharedPtr createChildManager(Event::Dispatcher& dispatcher) PURE;
```
I believe we want both of these methods. For listeners, we can configure the drain-type separately but for anything below the listener (such as filter-chains), we want to inherit that value.

antoniovicente commented:
Let me know when this PR is ready for another look.

murray-stripe commented:
Thanks for the additional feedback. I should have something ready near the end of this week.

murray-stripe commented on source/server/drain_manager_impl.cc:
```diff
@@ -52,15 +81,52 @@ bool DrainManagerImpl::drainClose() const {
+  const MonotonicTime current_time = server_.dispatcher().timeSource().monotonicTime();
+  const auto remaining_time =
+      std::chrono::duration_cast<std::chrono::seconds>(drain_deadline_ - current_time);
+  ASSERT(server_.options().drainTime() >= remaining_time);
```
Fixed.

murray-stripe commented:
I've included test fixes for everything that was failing in the last round, merged with `main` to pick-up the include changes and resolve conflicts, and addressed a bit of your feedback. 

Mainly, I updated the filter-chain to use TLS to create child-managers that get handed out for filter creation (via calls to `drainDecision()`). This eliminates a big portion of cross-thread interaction as callbacks are now registered with a drain-manager on their same thread. The only cross-thread interactions happen between drain-managers and their children. 

I'm considering now refactoring how children are managed and separating them from the general callback mechanism such that I can:
  + [ ] Make the callback mechanism _non-thread-safe_ (and assert registrations happen on the same thread)
  + [ ] Resolve the `child_drain_cbs_` growth problem

This should reduce a lot of the locking and simplify the API (no need to pass in dispatchers when registering callbacks). That should wrap up the unaddressed feedback and I'll be working on this refactor today and updating any new tests that may be failing. I'm hoping to wrap this work up by EOD or early tomorrow.

moderation commented:
Was just typing a question about the patch which triggered the dependency approval but looks like you just removed it :+1:

murray-stripe commented:
@antoniovicente I believe I have addressed all of the feedback as well as the performance concerns regarding cross-thread locking/coordination. I'll be watching for anymore tests in CI that need to be addressed, but I have finished adding additional test coverage for the drain-manager behaviors and am not planning on adding any additional tests at the moment.

I would appreciate your review.

antoniovicente commented on source/server/drain_manager_impl.h:
```diff
@@ -22,22 +27,39 @@ namespace Server {
   Event::TimerPtr drain_tick_timer_;
   MonotonicTime drain_deadline_;
+  mutable Common::CallbackManager<std::chrono::milliseconds> cbs_{};
+  std::vector<Common::ThreadSafeCallbackHandlePtr> child_drain_cbs_;
```
child_drain_cbs_ seems unused.

murray-stripe replied:
removed.

antoniovicente commented on source/server/filter_chain_manager_impl.cc:
```diff
@@ -28,14 +32,28 @@ Network::Address::InstanceConstSharedPtr fakeAddress() {
-    : parent_context_(parent_context), init_manager_(init_manager) {}
+    Configuration::DrainableFactoryContext& parent_context, Init::Manager& init_manager)
+    : parent_context_(parent_context), init_manager_(init_manager),
+      drain_manager_(parent_context.drainManager().createChildManager(parent_context.dispatcher())),
```
Do we need a call to createChildManager here, or could we use parent_context.drainManager() directly?

murray-stripe replied:
I think we could use the `parent_context.drainManager()` _today_ given the code, but if future code wanted to take advantage of the callback interface, then we'd want it to stay the way it is for correctness.

antoniovicente commented on source/server/listener_impl.cc:
```diff
@@ -415,11 +418,11 @@ void ListenerImpl::buildUdpListenerFactory(Network::Socket::Type socket_type,
+    // causes non-QUIC integration tests to fail, which I haven't investigated yet.
+    // Additionally, from looking at the GSO code there are substantial copying inefficiency so
+    // I don't think it's wise to enable to globally for now. I will circle back and fix both of
+    // the above with a non-QUICHE GSO implementation.
```
Do you know the reason behind these comment changes?  It seems like your formatter is doing something that it shouldn't be doing.

murray-stripe replied:
Hmm.. I use the `.devcontainer` (VS-Code) setup Envoy ships with (without any customizations). I can undo this formatting if it's a concern, but I'm guessing something about the standard tooling has changed since this file was last edited?

antoniovicente replied:
It would be good to know what caused these diffs.  Part of my reason for asking is that the multiple comment-only whitespace diffs I noticed in this PR make a very large PR even bigger and there are many of these whitespace only changes.

antoniovicente commented on source/server/drain_manager_impl.cc:
```diff
@@ -52,15 +80,74 @@ bool DrainManagerImpl::drainClose() const {
+    // If we've already passed our deadline, or we have an immediate drain-strategy, call the
+    // callback with no delay specified
+    if (current_time > drain_deadline_ ||
+        server_.options().drainStrategy() == Server::DrainStrategy::Immediate) {
```
Consider factoring this logic into the computation of remaining_time.  Something like:

```
std::chrono::milliseconds remaining_time =
  (server_.options().drainStrategy() != Server::DrainStrategy::Immediate && current_time < drain_deadline_) ?
  std::chrono::duration_cast<std::chrono::milliseconds>(drain_deadline_ - current_time)  :
  std::chrono::milliseconds{0};
cb(remaining_time);
return nullptr;
```

murray-stripe replied:
done!

antoniovicente commented on source/server/config_validation/server.cc:
```diff
@@ -53,7 +54,9 @@ ValidationInstance::ValidationInstance(
       http_context_(stats_store_.symbolTable()), router_context_(stats_store_.symbolTable()),
-      time_system_(time_system), server_contexts_(*this) {
+      time_system_(time_system), server_contexts_(*this),
+      drain_manager_(std::make_unique<Server::DrainManagerImpl>(
```
What methods are ever called on this top-level drain manager owned by the server?  It seems to me that drains are initiated at the listener or filter chain levels.  Is this layer of managers needed or could the parent DrainManagerImpl be created directly by source/server/listener_impl.cc prior to this PR?

murray-stripe replied:
There is a top-level drain-manager on the InstanceImpl that will initiate a server-wide drain-sequence via `drainListeners()`. Prior to this PR, this was the main drain-manager object. The validation server doesn't need a drain-manager since draining isn't necessarily something that it does.

This change is so that when it is constructing listeners and filter-chains, the child-drain manager creation bits don't fail where previously they were attempting to call the `drainManager()` function and hitting the `NOT_IMPLEMENTED_GCOVR_EXCL_LINE`.

antoniovicente reviewed (commented):
I'm having trouble following all the changes in this PR given its size, number of intermediate commits and comments that are still open.  It's still not clear to me why the DrainManagerImpl needs to be hierarchical and have the ability to invoke callbacks directly and forward the drain info to its children.

/wait

murray-stripe commented:
@antoniovicente Thank you for your reviews and I really appreciate your time.

> I'm having trouble following all the changes in this PR given its size, number of intermediate commits and comments that are still open.

Please let me know how I can approach this better and I'll do my best to make things easier on you.

> It's still not clear to me why the DrainManagerImpl needs to be hierarchical and have the ability to invoke callbacks directly and forward the drain info to its children.

Making the `DrainManagerImpl` hierarchical was to retain the current behaviors of the pull-based model using the `DrainDecision` interface. In effect, draining was already heirarchical in that if the main drain-manager was draining, listeners and filter-chains would also drain, listeners could drain as well as their filter-chains, and filter-chains could drain independent of listeners.

The major difference is that with a pull-based model (reading the drain-state via the `bool drainClose()` function), this was achieved by implementing the `DrainDecision` interface in various places which would allow the class to perform separate draining by checking a local state, and then the state of it's parent `DrainDecision` object.

Having the drain-manager invoke callbacks directly is the proposed approach #14350 and was the only proposed solution for changing this workflow to a push-based model. Pushing state to the children is just the necessary plumbing to retain the existing hierarchical drain behaviors in the push-based world.

murray-stripe commented on source/server/drain_manager_impl.cc:
```diff
@@ -14,7 +16,33 @@ namespace Server {
+
+  // Wire up the child so that when the parent starts draining, the child also sees the state-change
+  auto child_cb = children_.add(dispatcher, [child_ = std::weak_ptr<DrainManagerImpl>(child)] {
+    auto child = child_.lock();
```
The callback handle (returned by `children_.add(...)` is stored on `child` (near bottom of this function), so when the child drain-manager is destructed, it should remove itself from `children_` as well.

murray-stripe commented on source/server/api_listener_impl.h:
```diff
@@ -44,6 +45,7 @@ class ApiListenerImplBase : public ApiListener,
   // Network::DrainDecision
   // TODO(junr03): hook up draining to listener state management.
   bool drainClose() const override { return false; }
+  Common::CallbackHandlePtr addOnDrainCloseCb(DrainCloseCb) const override { return nullptr; }
```
It's not that it should never be called, but that the `ApiListenerImplBase` doesn't support draining (indicated by the `drainClose()` function always returning false.

I don't think it's correct to assume that it shouldn't be called, but the implementation of when it is called could be one of two options:

  0. Keep the current implementation (return `false` and `nullptr` for the `DrainDecision` interfaces)
  1. Create a local drain-manager object and proxy calls to it. Effectively the same result, but real callback handles are returned. 

The `TODO` comment above the `drainClose` section seems like there may be future work in this area and it seems out of scope for this change to address.

mattklein123 replied:
Please add more comments.

murray-stripe commented on source/server/drain_manager_impl.cc:
```diff
@@ -14,7 +16,33 @@ namespace Server {
 DrainManagerImpl::DrainManagerImpl(Instance& server,
                                    envoy::config::listener::v3::Listener::DrainType drain_type)
-    : server_(server), drain_type_(drain_type) {}
+    : server_(server), dispatcher_(server.dispatcher()), drain_type_(drain_type),
```
Removed.

murray-stripe commented on source/server/drain_manager_impl.cc:
```diff
@@ -52,15 +80,74 @@ bool DrainManagerImpl::drainClose() const {
+  drain_deadline_ = dispatcher_.timeSource().monotonicTime() + drain_delay;
+
+  // Call registered on-drain callbacks - immediately
+  if (server_.options().drainStrategy() == Server::DrainStrategy::Immediate) {
```
Done!

antoniovicente commented on source/server/listener_impl.h:
```diff
@@ -156,9 +158,9 @@ class PerListenerFactoryContextImpl : public Configuration::ListenerFactoryConte
+                                ListenerImpl& listener_impl, DrainManagerSharedPtr drain_manager)
       : listener_factory_context_base_(std::make_shared<ListenerFactoryContextBaseImpl>(
-            server, validation_visitor, config_message, std::move(drain_manager))),
+            server, validation_visitor, config_message, drain_manager)),
```
I think std::move was still appropriate here, it avoids changes to the shared_ptr refcount

antoniovicente commented on test/common/common/callback_impl_test.cc:
```diff
@@ -8,7 +12,7 @@ using testing::InSequence;
 namespace Common {
 
-class CallbackManagerTest : public testing::Test {
+class CallbackManagerTest : public testing::TestWithParam<bool> {
```
I don't think you're using the `TestWithParam<bool>` since I see no instantiations of the parametrized test class, calls to GetParam() or use of TEST_P.

Change back to `Test`

murray-stripe replied:
resolved in new PR.

antoniovicente commented on source/common/common/callback_impl.h:
```diff
@@ -81,5 +104,135 @@ template <typename... CallbackArgs> class CallbackManager {
+      });
+    }
+
+    Callback cb_;
```
I think this data member is unused.

murray-stripe replied:
This is used when calling the function in `runCallbacksWith`.

antoniovicente reviewed (commented):
> @antoniovicente Thank you for your reviews and I really appreciate your time.
> 
> > I'm having trouble following all the changes in this PR given its size, number of intermediate commits and comments that are still open.
> 
> Please let me know how I can approach this better and I'll do my best to make things easier on you.

I really don't know what to suggest.  I know I just spent several hours doing a review pass of the PR today, I don't have the time to do another pass like this.  This PR is very large, but I don't know what to suggest to split it.  I guess that the thread safe manager and the basic no-op plumbing could be separated from changes that take advantage of the new hooks. Minimizing/reverting whitespace changes in comments could help a bit too. One question I may need to ask at some point is about risk of this change and if it would be appropriate to runtime guard the changes.  So far I have gotten lost in the plumbing of the DrainManagerImpl part and thus have had a harder time understanding the functional changes that come from this PR.  I'm worried about the complexity of ThreadSafeCallbackManager object lifetimes which combine use of weak_ptr, cross thread interactions, handles, deferred removals via dispatcher post callbacks, etc.

> > It's still not clear to me why the DrainManagerImpl needs to be hierarchical and have the ability to invoke callbacks directly and forward the drain info to its children.
> 
> Making the `DrainManagerImpl` hierarchical was to retain the current behaviors of the pull-based model using the `DrainDecision` interface. In effect, draining was already heirarchical in that if the main drain-manager was draining, listeners and filter-chains would also drain, listeners could drain as well as their filter-chains, and filter-chains could drain independent of listeners.
> 
> The major difference is that with a pull-based model (reading the drain-state via the `bool drainClose()` function), this was achieved by implementing the `DrainDecision` interface in various places which would allow the class to perform separate draining by checking a local state, and then the state of it's parent `DrainDecision` object.
> 
> Having the drain-manager invoke callbacks directly is the proposed approach #14350 and was the only proposed solution for changing this workflow to a push-based model. Pushing state to the children is just the necessary plumbing to retain the existing hierarchical drain behaviors in the push-based world.

murray-stripe commented:
@antoniovicente I appreciate your time taken on these reviews. I'll work to split out the callback manager and additional API additions to the drain-manager into a separate PR. I'm not sure if there is another reviewer who could pick up the new PR to help reduce your load?

Once that PR merges, I'll update this branch to be much simplified.

murray-stripe commented on source/common/common/callback_impl.h:
```diff
@@ -81,5 +104,135 @@ template <typename... CallbackArgs> class CallbackManager {
+   * @brief Run all callbacks with a function that returns the input arguments
+   *
+   * NOTE: This code is currently safe if a callback deletes ITSELF from within a callback. It is
+   *       not safe if a callback deletes other callbacks.
```
This has been removed in the new PR.

murray-stripe commented on source/common/common/callback_impl.h:
```diff
@@ -81,5 +104,135 @@ template <typename... CallbackArgs> class CallbackManager {
+   * executed for each callback. Return values must be able to be safely copied across threads and
+   * asynchronously consumed.
+   */
+  void runCallbacksWith(std::function<std::tuple<CallbackArgs...>(void)> run_with) {
```
I have updated the class to be non-templatized for now as it's not needed. And an argument could be made that passing around input arguments across thread may be a complexity that we don't want to deal with. I've simplified this in the new PR.

murray-stripe commented on test/common/common/callback_impl_test.cc:
```diff
@@ -53,5 +57,82 @@ TEST_F(CallbackManagerTest, DestroyManagerBeforeHandle) {
+  ON_CALL(dispatcher, post(_)).WillByDefault(Invoke([](std::function<void()> cb) { cb(); }));
+  ON_CALL(cb_dispatcher, post(_)).WillByDefault(Invoke([](std::function<void()> cb) { cb(); }));
+
+  ThreadSafeCallbackManager<int> manager{dispatcher};
```
I've removed the templatized version and all tests now reflect a zero-arg version. In new PR.

murray-stripe commented:
@antoniovicente I have opened #17026 which includes the API changes to `DrainDecision` and the `DrainManagerImpl` implementation, the `ThreadSafeCallbackManager`, the additional tests for both classes, and the necessary related changes to other tests/mocks/classes to address the APIs.

The changes in the PR are all no-op. I'll focus my efforts on getting that PR into final shape and merging before coming back to this PR for a major cleanup to reduce the amount of change in this file. 

Again, thank you very much for your time and reviews, they are greatly appreciated!

murray-stripe commented on source/server/filter_chain_manager_impl.h:
```diff
@@ -78,12 +82,15 @@ class PerFilterChainFactoryContextImpl : public Configuration::FilterChainFactor
 
-  void startDraining() override { is_draining_.store(true); }
+  void startDraining() override {
+    drain_manager_->startDrainSequence([] {});
```
Addressed [here](https://github.com/envoyproxy/envoy/pull/17026#issuecomment-863531127).

murray-stripe commented:
@antoniovicente I'm working on updating this PR now that the preamble PR has merged (#17026). I'll remove the "WIP" from the PR and mention you when this is ready for re-review.

murray-stripe commented:
/retest

murray-stripe commented:
ah, didn't realize CI merged `main` before running tests. Was confused for a while why the line-numbers weren't lining up in the compile error. :-D 

@antoniovicente This should be ready for your review again. I'll be keeping an eye on CI, but outside of errors from merging with main in the tests, I don't expect any issues.

murray-stripe commented:
Macos timeout
/retest

JohnMurray commented:
@antoniovicente Just following up as a friendly reminder that this is ready for review.

(Same murray, just not access to my work machine at the moment).

antoniovicente commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -484,6 +504,25 @@ void ConnectionManagerImpl::doConnectionClose(
+      stats_.named_.downstream_cx_drain_close_.inc();
+      for (const auto& stream : streams_) {
+        ENVOY_STREAM_LOG(debug, "drain closing connection", *stream);
+      }
```
I'm not sure why this logging is here instead of startDrainSequence.  What makes the start of a drain timer different from other calls to startDrainSequence?

antoniovicente commented on test/server/listener_manager_impl_test.cc:
```diff
@@ -1700,7 +1700,7 @@ reuse_port: false
 
 TEST_F(ListenerManagerImplTest, UpdateBindToPortEqualToFalse) {
-  InSequence s;
+  // InSequence s;
```
Either restore or remove commented out code.

murray-stripe replied:
Uncommented.

antoniovicente commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -484,6 +504,25 @@ void ConnectionManagerImpl::doConnectionClose(
+    if (drain_state_ != DrainState::NotDraining) {
+      return;
+    }
+    if (!codec_) {
```
Should createStartDrainTimer ever be called when codec is no longer present?  Should we be creating timers for drain after doing the doConnectionClose a few lines down?

I think there's some inconsistencies in implementation that add to complexity.  Would be good to avoid a call to createStartDrainTimer once the codec has been deleted, and better understand cases where this situation currently happens.

murray-stripe replied:
I've moved the `doConnectionClose` outside of the timer creation. My original implementation hadn't fully considered the lifespan of the `codec_`. Reading through the code I believe you're right that there isn't any reason to wrap the logic in a timer if the `codec_` is null when creating the timer.

murray-stripe commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -154,6 +155,16 @@ void ConnectionManagerImpl::initializeReadFilterCallbacks(Network::ReadFilterCal
+
+  // register callback for drain-close events
+  start_drain_cb_ =
+      drain_close_.addOnDrainCloseCb([this](std::chrono::milliseconds drain_delay) -> void {
```
Removed.

murray-stripe commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -484,6 +504,25 @@ void ConnectionManagerImpl::doConnectionClose(
+      doConnectionClose(Network::ConnectionCloseType::FlushWrite, absl::nullopt, "");
+    } else {
+      startDrainSequence();
+      stats_.named_.downstream_cx_drain_close_.inc();
```
Ah, this was intended to be the same stat. I've kept it in two places, but only because the `doConnectionClose` call was moved outside of the timer.

murray-stripe commented:
Thank you for your feedback @antoniovicente. I've addressed your feedback and I've also added a feature flag (which we talked about in a separate comment at some point in this PR or the prerequisite one). 

As part of the feature-flag I also re-added the previous logic in the encodeHeaders function to send the GOAWAY. I noticed that for the Admin handler, trying to create a new ConnectionManagerImpl already in a draining state wouldn't allow for connections to be established and that the Admin handler had been relying on the encodeHeaders approach.

antoniovicente commented on source/common/http/conn_manager_impl.h:
```diff
@@ -67,7 +67,8 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
                         Upstream::ClusterManager& cluster_manager,
-                        Server::OverloadManager& overload_manager, TimeSource& time_system);
+                        Server::OverloadManager& overload_manager, TimeSource& time_system,
+                        bool use_proactive_draining = true);
```
Can you comment on use of an argument instead of reading the envoy.reloadable_features.http_conn_manager_proactive_drain runtime feature inside the constructor?  Reading the runtime feature in the constructor seems more consistent.

murray-stripe replied:
By admin handlers I'm referring to the set of HTTP endpoints on Envoy's admin interface (`/config_dump`, `/stats`, etc).

antoniovicente commented on source/common/http/conn_manager_impl.h:
```diff
@@ -461,6 +465,7 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
   bool clear_hop_by_hop_response_headers_{true};
   // The number of requests accumulated on the current connection.
   uint64_t accumulated_requests_{};
+  bool use_proactive_draining_;
```
nit: const

murray-stripe replied:
done in https://github.com/envoyproxy/envoy/pull/16201/commits/80aed8cbe5b9a3132befdbcdc3ffb8fe865f3f86

antoniovicente commented on source/server/admin/admin.cc:
```diff
@@ -240,7 +240,7 @@ bool AdminImpl::createNetworkFilterChain(Network::Connection& connection,
       *this, server_.drainManager(), server_.api().randomGenerator(), server_.httpContext(),
       server_.runtime(), server_.localInfo(), server_.clusterManager(), null_overload_manager_,
-      server_.timeSource())});
+      server_.timeSource(), /*use_proactive_draining=*/false)});
```
Disabling of draining on the admin interface means that we need to keep 2 code paths through the HCM indefinitely.  Is this divergence in behavior needed?

murray-stripe replied:
It is at least needed for now as the admin interfaces assume they can always open new connections after entering a draining state.

I don't fully understand the use-cases for draining the admin interfaces. I would assume if this is the behavior that we want, that we don't need to drain these interfaces at all. If that were to be the case, then we could simply pass a new drain-manager instance (not connected to the server drain manager) that would never initiate draining.

murray-stripe replied:
I've added a TODO. I can follow-up in another PR to ensure that the admin server does not drain.

antoniovicente reviewed (approved):
Change looks good as far as I can tell.

/assign-from @envoyproxy/senior-maintainers

murray-stripe commented:
@lizan - Just a friendly ping that this is ready for review.

lizan commented on test/common/http/conn_manager_impl_test_base.h:
```diff
@@ -170,6 +174,9 @@ class HttpConnectionManagerImplTest : public testing::Test, public ConnectionMan
   ConnectionManagerTracingStats tracing_stats_{CONN_MAN_TRACING_STATS(POOL_COUNTER(fake_stats_))};
+  Event::MockTimer* drain_begin_timer_;
+  Network::DrainDecision::DrainCloseCb drain_close_func_;
+  bool drain_close_func_set_{false};
```
nit: do we need this? You can check if drain_close_func_ == nullptr?

murray-stripe replied:
Good point. Updated.

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -500,6 +522,23 @@ void ConnectionManagerImpl::doConnectionClose(
+    if (drain_state_ != DrainState::NotDraining) {
+      return;
+    }
+    stats_.named_.downstream_cx_drain_close_.inc();
```
I would move this into the `startDrainSequence()` function and then you consolidate this stat increment.

mattklein123 commented on source/server/filter_chain_manager_impl.cc:
```diff
@@ -1,6 +1,10 @@
 #include "source/server/filter_chain_manager_impl.h"
 
 #include "envoy/config/listener/v3/listener_components.pb.h"
+#include "envoy/event/dispatcher.h"
```
Given the risk of the HCM change, I would really rather you split this part of the change into a separate PR, but I'm not going to demand it if you don't feel that it's worth it.

murray-stripe replied:
Responded in top-level comment [here](https://github.com/envoyproxy/envoy/pull/16201#pullrequestreview-750767228)

mattklein123 commented on source/server/listener_impl.h:
```diff
@@ -148,7 +146,7 @@ class ListenerFactoryContextBaseImpl final : public Configuration::FactoryContex
   Stats::ScopePtr global_scope_;
   Stats::ScopePtr listener_scope_; // Stats with listener named scope.
   ProtobufMessage::ValidationVisitor& validation_visitor_;
-  const Server::DrainManagerPtr drain_manager_;
```
I think this can still be const?

mattklein123 reviewed (changes requested):
Thanks this LGTM at a high level. Flushing out some comments. As I mentioned in one of the comments my preference would be to split out the factory changes from the HCM changes since the HCM changes are very risky and I would rather test the factory changes in isolation which AIUI should still work fine with the existing code. Thank you.

/wait

murray-stripe commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -141,6 +142,17 @@ void ConnectionManagerImpl::initializeReadFilterCallbacks(Network::ReadFilterCal
+  // register callback for drain-close events
+  start_drain_cb_ =
+      drain_close_.addOnDrainCloseCb(read_callbacks_->connection().dispatcher(),
+                                     [this](std::chrono::milliseconds drain_delay) -> void {
```
Done.

murray-stripe commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -471,6 +484,15 @@ void ConnectionManagerImpl::doConnectionClose(
 
+  if (start_drain_cb_) {
+    start_drain_cb_.reset();
+  }
```
Removed.

murray-stripe commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1168,6 +1207,11 @@ void ConnectionManagerImpl::ActiveStream::disarmRequestTimeout() {
+
+  for (const auto& stream : streams_) {
+    ENVOY_STREAM_LOG(debug, "drain closing connection", *stream);
+  }
```
Cleaned up to use a connection log.

murray-stripe reviewed (commented):
I believe I've addressed the bulk of the feedback (sorry for commit-churn while I tried to work around the DCO issues). I'll work on writing documentation next and any other feedback that I may have missed.

Regarding the split of functionality, I don't have strong opinions against it. I don't believe there is a ton of added value in splitting unless we're talking about splitting the change across versions (I guess I'm skeptical that we'd get enough validation to make it worth it, but perhaps there are other arguments to be made). The behavior within the HCM is feature-flagged, so it seems like the isolation controls are already in place.

I'll ping directly when this is ready for review.

murray-stripe commented on api/envoy/admin/v3/server_info.proto:
```diff
@@ -80,7 +80,8 @@ message CommandLineOptions {
   }
 
   enum DrainStrategy {
-    // Gradually discourage connections over the course of the drain period.
```
I've taken an initial stab at updating the documentation.

murray-stripe commented:
/retest

mattklein123 commented:
@murray-stripe sorry I've lost the thread on this. Are you ready to have this looked at again? Can you fix CI?

/wait

murray-stripe commented:
/retest

murray-stripe commented:
/retest

@mattklein123  This change is ready for re-review. Just currently dealing with a test timeout that I can repro locally and assuming it's a flake.

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1392,7 +1431,6 @@ void ConnectionManagerImpl::ActiveStream::encodeHeaders(ResponseHeaderMap& heade
     // prevent any new streams.
     connection_manager_.startDrainSequence();
     connection_manager_.stats_.named_.downstream_cx_drain_close_.inc();
-    ENVOY_STREAM_LOG(debug, "drain closing connection", *this);
```
Why was this removed but the stat wasn't? related to my comment above about possible double stat increments.

mattklein123 reviewed (changes requested):
Thanks generally LGTM. Flushing out some more comments.

This change is sufficiently scary that I would appreciate a smoke test. Can you please deploy this change at Stripe in some test capacity to make sure there are no obvious crashes/issues?

/wait

murray-stripe commented:
Haven't forgotten about this PR, but will have to circle back around in Q1 due to internal reprioritization in order to provide a sufficient "bake" in Stripe's environment. I have done some initial testing, but would like to do some more sustained testing to provide better assurance of the safety of the change.
